### DISCUSSION TITLE
REST API: How to send a `POST` request with a query and a subquery?

### DISCUSSION THREAD

**Discussion by javac9** [opencrx/opencrx-documentation#22] [Q&A]
Return all `Contact`s that satisfy the following

**requirements:**
- feature `firstName` begins with letter `S`
- has a `PostalAddress` with a `PostalCountry` code of `240`

The first requirement is achieved through the following query:
```
{
    "org.openmdx.kernel.Query": {
        "queryType": "org:opencrx:kernel:account1:Contact",
        "queryFilter": {
            "condition": {
                "@type": "org:openmdx:kernel:Condition",
                "objects": [
                    {
                        "feature": "firstName",
                        "quantifier": 2,
                        "type": 5,
                        "value": {
                            "_item": [
                                {
                                    "$": "S.*"
                                }
                            ]
                        }
                    }
                ]
            }
        }
    }
}
```
The second requirement, if the request is sent separately to `http://localhost:9090/opencrx-rest-CRX/org.opencrx.kernel.account1/provider/CRX/segment/Standard/address`:
```
{
    "org.openmdx.kernel.Query": {
        "queryType": "org:opencrx:kernel:account1:PostalAddress",
        "queryFilter": {
            "condition": {
                "objects": [
                    {
                        "quantifier": 2,
                        "type": 5,
                        "feature": "postalCountry",
                        "value": {
                            "_item": [
                                {
                                    "$": "240"
                                }
                            ]
                        }
                    }
                ]
            }
        }
    }
}
```
The real issue is how to combine them together? In Java API, this is done like this:
`contactQuery.thereExistsAddress().elementOf(PersistenceHelper.asSubquery(postalAddressQuery))`

By looking at the `Example value | Model` on the Swagger page, I can't see anything that could potentially serve the purpose of the above `elementOf(subquery)` _**query-connecting**_ method.

**Comment by javac9** [comment]
Could you provide any feedback - I suppose this should be possible with OpenCRX RESTful API?
I can try to provide more details if you need them..

**Comment by wfro64** [comment]
Please have a look at https://github.com/opencrx/opencrx-documentation/blob/master/Sdk/Rest.md#debugging. The simplest way to construct a complex JSON request body is as follows:

* Create a native Java client
* Run the Java client with a proxy server and monitor the network traffic
* Create a JSON template based on the network traffic and use it in your REST client. 

[TestQuery](https://github.com/opencrx/opencrx/blob/master/core/src/test/java/org/opencrx/kernel/api/TestQuery.java) contains many examples for nested queries.
